pkg/openstackipset/funcs.go

```diff
@@ -0,0 +1,244 @@
+	r common.ReconcilerCommon,
+	obj client.Object,
+	cond *ospdirectorv1beta1.Condition,
+	name string,
+	networks []string,
+	hostCount int,
+	vip bool,
+	serviceVIP bool,
+	deletedHosts []string,
+	addToPredictableIPs bool,
+) (map[string]ospdirectorv1beta1.HostStatus, reconcile.Result, error) {
+
+	status := map[string]ospdirectorv1beta1.HostStatus{}
+
+	//
+	// create IPSet to request IPs for all networks
+	//
+	_, err := createOrUpdateIPSet(
+		r,
+		obj,
+		cond,
+		name,
+		networks,
+		hostCount,
+		vip,
+		serviceVIP,
+		deletedHosts,
+		addToPredictableIPs,
+	)
+	if err != nil {
+		return status, reconcile.Result{}, err
+	}
+
+	//
+	// get ipset and verify all IPs got created
+	//
+	ipSet := &ospdirectorv1beta1.OpenStackIPSet{}
+	err = r.GetClient().Get(context.TODO(), types.NamespacedName{
+		Name:      strings.ToLower(name),
+		Namespace: obj.GetNamespace()},
+		ipSet)
+	if err != nil {
+		cond.Message = fmt.Sprintf("Failed to get %s %s ", ipSet.Kind, ipSet.Name)
+		cond.Reason = ospdirectorv1beta1.ConditionReason(ospdirectorv1beta1.IPSetCondReasonError)
+		cond.Type = ospdirectorv1beta1.ConditionType(ospdirectorv1beta1.CommonCondTypeError)
+		err = common.WrapErrorForObject(cond.Message, obj, err)
+
+		return status, reconcile.Result{}, err
+	}
+
+	//
+	// check for all hosts got created on the IPset
+	//
+	if len(ipSet.Status.Hosts) < hostCount {
+		cond.Message = fmt.Sprintf("Waiting on hosts to be created on IPSet %v - %v", len(ipSet.Status.Hosts), hostCount)
+		cond.Reason = ospdirectorv1beta1.ConditionReason(ospdirectorv1beta1.IPSetCondReasonWaitingOnHosts)
+		cond.Type = ospdirectorv1beta1.ConditionType(ospdirectorv1beta1.CommonCondTypeWaiting)
+
+		return status, reconcile.Result{RequeueAfter: 10 * time.Second}, nil
+	}
+
+	//
+	// check if hosts got deleted
+	//
+	if len(ipSet.Status.Hosts) > hostCount &&
+		(len(ipSet.Status.Hosts)-len(deletedHosts)) == hostCount {
+		for _, hostname := range deletedHosts {
+			delete(ipSet.Status.Hosts, hostname)
+		}
+
+	}
+
+	//
+	// get OSNetCfg object
+	//
+	osnetcfg := &ospdirectorv1beta1.OpenStackNetConfig{}
+	err = r.GetClient().Get(context.TODO(), types.NamespacedName{
+		Name:      strings.ToLower(obj.GetLabels()[openstacknetconfig.OpenStackNetConfigReconcileLabel]),
+		Namespace: obj.GetNamespace()},
+		osnetcfg)
+	if err != nil {
+		cond.Message = fmt.Sprintf("Failed to get %s %s ", osnetcfg.Kind, osnetcfg.Name)
+		cond.Reason = ospdirectorv1beta1.ConditionReason(ospdirectorv1beta1.MACCondReasonError)
+		cond.Type = ospdirectorv1beta1.ConditionType(ospdirectorv1beta1.CommonCondTypeError)
+		err = common.WrapErrorForObject(cond.Message, obj, err)
+
+		return status, reconcile.Result{}, err
+	}
+
+	for hostname, hostStatus := range ipSet.Status.Hosts {
+		err = openstacknetconfig.WaitOnIPsCreated(
+			r,
+			obj,
+			cond,
+			osnetcfg,
+			networks,
+			hostname,
+			&hostStatus,
+		)
+		if err != nil {
+			return status, reconcile.Result{RequeueAfter: 10 * time.Second}, nil
```

Comment:
Do we want to return `err` here since it's non-nil?

Reply:
I guess `openstacknetconfig.WaitOnIPsCreated` only ever returns "not found" errors, so perhaps it's unimportant.

Reply:
I wanted the controller just reconcile with the info message and not log an error. thats why I return a non default ctrl.Result and nil. if we ever return another error that would get logged as an error.